Catch malformed error request

riiwo opened the pull request:
In case of AWS service responds with malformed XML, erlang exits. We want to
catch that to identify the root cause of the problem

riiwo committed:
Catch malformed error request

In case of AWS service responds with malformed XML, erlang exits. We want to
catch that to identify the root cause of the problem by logging it out. Keeping
behaviour by exiting as before.

indrekj commented on lib/swoosh/adapters/amazon_ses.ex:
```diff
@@ -104,12 +104,17 @@ defmodule Swoosh.Adapters.AmazonSES do
+      %{code: code, message: message}
+    catch
+      # In case of malformed XML erlang is not throwing error, but is exiting instead
+      :exit, _ -> %{code: 400, message: body}
```
I think it would be a bit better if we log out the error and the body, but still re-raise the error. So we know what to fix but don't actually suppress the actual error.

indrekj commented on lib/swoosh/adapters/amazon_ses.ex:
```diff
@@ -104,12 +104,17 @@ defmodule Swoosh.Adapters.AmazonSES do
+      %{code: code, message: message}
+    catch
+      # In case of malformed XML erlang is not throwing error, but is exiting instead
+      :exit, error -> %{code: 400, message: error, malformed_body: body}
```
what about:
```
    rescue
      exception ->
         Logger.error(...)
         reraise exception, __STACKTRACE__
```
I would do this to keep the behaviour the same and just add logging. I guess your option is also fine but you likely then have to change the service to actually log these out and make sure we don't suppress the errors

indrekj reviewed (approved):
i guess `rescue` didn't work. you could also test the logging part which is the most useful part here for us but I'm fine with this as well.

riiwo commented:
> i guess `rescue` didn't work. you could also test the logging part which is the most useful part here for us but I'm fine with this as well.

Added test

riiwo committed:
Rescue error when error response from SES is empty

We don't want to get fatal when error response is empty as it usually recovers itself.

Code for response was chosen from:
https://docs.aws.amazon.com/ses/latest/DeveloperGuide/using-ses-api-error-codes.html

riiwo commented:
@indrekj  added another commit for empty responses

indrekj commented on lib/swoosh/adapters/amazon_ses.ex:
```diff
@@ -90,7 +90,12 @@ defmodule Swoosh.Adapters.AmazonSES do
-        {:error, parse_error_response(body)}
+        case body do
+          "" ->
+            {:error, %{code: "ServiceUnavailable", message: "Empty response from SES"}}
```
do you know what's the status code here? The `when code > 399` part indicates that it's at least 400 but what exactly?

riiwo replied:
It wasn't in the logs. I'll include it in the logging perhaps?

riiwo replied:
My best guess is that it times-out as there is no response, so 522? So maybe I should use timeout instead

indrekj replied:
In any case, let's log it out to be sure. You can add a `Logger...` statement before this line. It might be 503 or smth else as well. So later maybe we can just check the status code and not even try to parse the body if we don't need to. Then we can also do an upstream PR.

riiwo replied:
Added logging

indrekj commented on test/swoosh/adapters/amazonses_test.exs:
```diff
@@ -154,6 +164,46 @@ defmodule Swoosh.Adapters.AmazonSESTest do
+      {:error, %{code: "ServiceUnavailable", message: "Empty response from SES"}}
+
+    assert capture_log(fn ->
+      Logger.error("Empty response with code: 500")
```
This doesn't test the code. You're logging it out manually and then asserting the same thing. Instead, the capture_log should be around `AmazonSES.deliver` for it to be a real test.

riiwo replied:
Ok, maybe now 😄

indrekj replied:
That makes more sense now :+1: Lets deploy this and wait for the next occurrence. Then I think we can simplify it and make an upstream PR as well.

wifilinker committed:
Log malformed error responses from AWS SES

There are times when AWS SES seems to be temporarily unavailable - for
a limited amount of time it returns an error response for perfectly OK
emails. This gets solved on its own when we resend the same message at
a later time. The issue is that error handling fails when trying to read
the error code from the response.

Logging the error to understand what's happening.

bautrey37 commented:
Mail-courier was still using this version of swoosh. I did not see any of the logs added in this PR on Kibana, so I'm assuming that this issue does not exist anymore. I've [upgraded](https://github.com/salemove/mail-courier/commit/7c513ef03e5bd068300cf3ff0e463a1a13dffa7d) mail-courier to use the latest version of swoosh.
